model/task_queue_service.go

```diff
@@ -0,0 +1,197 @@
+	units    map[string]schedulableUnit
+}
+
+// schedulableUnit represents a unit of tasks that must be kept together in the queue because they
+// are pinned to some number of hosts. That is, it represents tasks and task groups, but it does
+// _not_ represent builds or DAGs of tasks and their dependencies.
+type schedulableUnit struct {
+	id           string
+	runningHosts int // number of hosts unit is currently running on
+	maxHosts     int // number of hosts unit can run on
+	tasks        []TaskQueueItem
+}
+
+// NewTaskDispatchService creates a TaskDispatchService from a slice of TaskQueueItems.
+func NewTaskDispatchService(distroID string, items []TaskQueueItem) *TaskDispatchService {
+	// This slice likely has too much capacity, but it helps append performance.
+	order := make([]string, 0, len(items))
+	units := map[string]schedulableUnit{}
+	var ok bool
+	var unit schedulableUnit
+	var id string
+
+	for _, item := range items {
+		if item.Group == "" {
+			order = append(order, item.Id)
+			units[item.Id] = schedulableUnit{
+				id:       item.Id,
+				maxHosts: 0, // maxHosts == 0 indicates not a task group
+				tasks:    []TaskQueueItem{item},
+			}
+		} else {
+			// If it's the first time encountering the task group, save it to the order
+			// and create an entry for it in the map. Otherwise, append to the
+			// TaskQueueItem array in the map.
+			id = compositeGroupId(item.Group, item.BuildVariant, item.Version)
+			if _, ok = units[id]; !ok {
+				units[id] = schedulableUnit{
+					id:       id,
+					maxHosts: item.GroupMaxHosts,
+					tasks:    []TaskQueueItem{item},
+				}
+				order = append(order, id)
+			} else {
+				unit = units[id]
+				unit.tasks = append(unit.tasks, item)
+				units[id] = unit
+			}
+		}
+	}
+	t := &TaskDispatchService{
+		distroID: distroID,
+		order:    order,
+		units:    units,
+	}
+	return t
+}
+
+// FindNextTask returns the next dispatchable task in the queue.
+func (t *TaskDispatchService) FindNextTask(spec TaskSpec) *TaskQueueItem {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	// If units is empty, the queue has been emptied. Reset order as an optimization so that the
+	// service no longer needs to iterate over it and check each item against the map.
+	if len(t.units) == 0 {
+		t.order = []string{}
+		return nil
+	}
+
+	var unit schedulableUnit
+	var ok bool
+	var next *TaskQueueItem
+
+	if spec.Group != "" {
+		unit, ok = t.units[compositeGroupId(spec.Group, spec.BuildVariant, spec.Version)]
+		if ok {
+			if next = t.nextTaskGroupTask(unit); next != nil {
+				return next
+			}
+		}
+		// If the task group is not present in the task group map, it has been dispatched.
+		// Fall through to getting a task not in that group.
+	}
+	var numHosts int
+	var err error
+	for _, schedulableUnitID := range t.order {
+		unit, ok = t.units[schedulableUnitID]
+		if !ok {
+			continue
+		}
+		// If maxHosts is not set, this is not a task group.
+		if unit.maxHosts == 0 {
+			delete(t.units, schedulableUnitID)
```

Comment:
Should we also remove it from the order?

Reply:
probably but that's really annoying and lines 109-112 make it less important

Reply:
Removing from the order is expensive. Removing an element from a long slice is cheap in Go only if you don't care about their order (since you can swap the middle element with an element at the end, and truncate the slice). Every remove from a slice requires allocating a new slice. Deleting from a map should be cheaper, and I check everywhere whether the thing is actually in the map. The map is the source of truth.

Reply:
Agreed. I think it's fine to leave this